tensornetwork/tn_keras/mpo.py

```diff
@@ -0,0 +1,167 @@
+import tensorflow as tf
+from tensorflow.keras.layers import Layer  # type: ignore
+from tensorflow.keras import activations
+from tensorflow.keras import initializers
+from typing import List, Optional, Text, Tuple
+import tensornetwork as tn
+from tensornetwork.network_components import Node
+import numpy as np
+import math
+
+
+class DenseMPO(Layer):
+  """Matrix Product Operator (MPO) TN layer.
+
+  Example:
+
+  ```python
+  # as first layer in a sequential model:
+  model = Sequential()
+  model.add(
+    DenseMPO(1024, num_nodes=4, bond_dim=8, activation='relu', input_dim=1024))
+  # now the model will take as input arrays of shape (*, 1024)
+  # and output arrays of shape (*, 1024)
+  # after the first layer, you don't need to specify
+  # the size of the input anymore:
+  model.add(DenseMPO(1024, num_nodes=4, bond_dim=8, activation='relu'))
+  ```
+
+  Args:
+    output_dim: Positive integer, dimensionality of the output space.
+    num_nodes: Positive integer, number of nodes in the MPO.
+      Note input_dim**(1. / num_nodes) and output_dim**(1. / num_nodes)
+      must both be round.
+    bond_dim: Positive integer, size of the intermediate dimension.
+    activation: Activation function to use.
+      If you don't specify anything, no activation is applied
+      (ie. "linear" activation: `a(x) = x`).
+    use_bias: Boolean, whether the layer uses a bias vector.
+    kernel_initializer: Initializer for the node weight matrices.
+    bias_initializer: Initializer for the bias vector.
+
+  Input shape:
+    2D tensor with shape: `(batch_size, input_dim)`.
+
+  Output shape:
+    2D tensor with shape: `(batch_size, output_dim)`.
+  """
+
+  def __init__(self,
+               output_dim: int,
+               num_nodes: int,
+               bond_dim: int,
+               use_bias: Optional[bool] = True,
+               activation: Optional[Text] = None,
+               kernel_initializer: Optional[Text] = 'glorot_uniform',
+               bias_initializer: Optional[Text] = 'zeros',
+               **kwargs) -> None:
+    if 'input_shape' not in kwargs and 'input_dim' in kwargs:
+      kwargs['input_shape'] = (kwargs.pop('input_dim'),)
```

Comment:
Allows the user to specify `input_dim` instead of `input_shape` - this can be done in other Keras layers like Dense and I wanted to make the layer creation as consistent as possible. Of course, this is redundant and we could just force the user to specify `input_shape=(input_dim,)` in cases where `input_dim` would otherwise be used. I'll keep it for now, and remove it if you feel strongly we should enforce a full `input_shape`.

Reply:
It's not necessary to handle this here - the superclass will do this for you.

Reply:
Edit: I see now. Yes, if we want to make these drop-in replacements for Dense, we should probably support this (but we shouldn't use it in the docstring, since it's a legacy modality). Please also put a comment describing that for posterity :)

Reply:
Great - updated the docstring and added a note.